Array append for optional element and sequences

jevonmao opened the pull request:
## PR Summary
### Motivation
The concept of optional types has been deeply ingrained into the design of Swift, and developers deal with appending optional elements to an array often.

It can be very inconvenient to be first unwrap optional, and append it:
```swift
var foo: [Double] = []
let bar: Double? = 69.0
if let unwrappedBar = bar {
   foo.append(unwrappedBar)
}
//or
if bar != nil {
   foo.append(bar!)
}
```

### Solution
By moving the logic of optional unwrap and checking to an `Array` extension method, developers can now use syntactical sugar:
```swift
var foo: [Double] = []
let bar: Double? = 69.0
foo.append(bar)
```

## Checklist
<!--- Please go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] I checked the [**Contributing Guidelines**](https://github.com/SwifterSwift/SwifterSwift/blob/master/CONTRIBUTING.md) before creating this request.
- [x] New extensions are written in Swift 5.0.
- [x] New extensions support iOS 10.0+ / tvOS 9.0+ / macOS 10.10+ / watchOS 2.0+, or use `@available` if not.
- [x] I have added tests for new extensions, and they passed.
- [x] All extensions have a **clear** comments explaining their functionality, all parameters and return type in English.
- [x] All extensions are declared as **public**.
- [x] I have added a [changelog](https://github.com/SwifterSwift/SwifterSwift/blob/master/CHANGELOG_GUIDELINES.md) entry describing my changes.

LucianoPAlmeida commented on Sources/SwifterSwift/SwiftStdlib/ArrayExtensions.swift:
```diff
@@ -45,6 +45,30 @@ public extension Array {
+     - Parameter newElements: The optional element to append to the array
+     */
+    @discardableResult
+    mutating func append(_ newElement: Self.Element?) -> [Element]
```
I think we wouldn't need to return the array here... just. avoid mutating func would be fine...

LucianoPAlmeida commented on Sources/SwifterSwift/SwiftStdlib/ArrayExtensions.swift:
```diff
@@ -45,6 +45,30 @@ public extension Array {
+     - Parameter newElements: The optional sequence to append to the array
+     */
+    @discardableResult
+    mutating func append<S>(contentsOf newElements: S?) -> [Element] where Element == S.Element, S : Sequence
```
Same here

LucianoPAlmeida reviewed (commented):
I think we could add this extension, but two things to think about are:
* Naming should be more suggestive? e.g. `appendIfNonNil`?
* This could be a `RangeReplaceableCollection`?

jevonmao commented:
@LucianoPAlmeida Thank you for the suggestions. I changed the name to `appendIfNotNil`, moved the extension to `RangeReplaceableCollection`, and removed the return statement.

LucianoPAlmeida commented on Sources/SwifterSwift/SwiftStdlib/RangeReplaceableCollectionExtensions.swift:
```diff
@@ -174,4 +174,24 @@ public extension RangeReplaceableCollection {
+     - Parameter newElements: The optional element to append to the array
+     */
+    mutating func appendIfNonNil(_ newElement: Self.Element?)
+        {
```
Nit: Formatting :)

LucianoPAlmeida commented on Sources/SwifterSwift/SwiftStdlib/RangeReplaceableCollectionExtensions.swift:
```diff
@@ -174,4 +174,24 @@ public extension RangeReplaceableCollection {
+     SwifterSwift: Adds a new element at the end of the array, mutates the array in place
+     - Parameter newElements: The optional element to append to the array
+     */
+    mutating func appendIfNonNil(_ newElement: Self.Element?)
```
Nit: 
```suggestion
    mutating func appendIfNonNil(_ newElement: Element?)
```

LucianoPAlmeida commented on Sources/SwifterSwift/SwiftStdlib/RangeReplaceableCollectionExtensions.swift:
```diff
@@ -174,4 +174,24 @@ public extension RangeReplaceableCollection {
+     - Parameter newElements: The optional sequence to append to the array
+     */
+    mutating func appendIfNonNil<S>(contentsOf newElements: S?) where Element == S.Element, S : Sequence
+        {
```
Nit: Braces start on the same line, this keeps consistent with the rest of the project =]

LucianoPAlmeida replied:
Also seems like there is one extra tab here ...

jevonmao commented:
Accidentally modified the entire file because I did reindent. I will fix that.

LucianoPAlmeida commented:
> The entire code file appears to be modified because I did reindent. Xcode removed extra tabs in the lines, is that good? I think SwiftLint will automatically fix it anyways if that's not desired.

Just run a `swiftlint autocorrect file.swift` and it would be fine =]

jevonmao commented:
Nothing came up with`swiftlint autocorrect`, so there is nothing for me to commit:)

LucianoPAlmeida commented on Sources/SwifterSwift/SwiftStdlib/RangeReplaceableCollectionExtensions.swift:
```diff
@@ -79,14 +79,14 @@ public extension RangeReplaceableCollection {
         return remove(at: randomIndex)
     }
-
+    
```
So, since nothing happen with auto correct, can you please remove those additions manually?

LucianoPAlmeida commented on Sources/SwifterSwift/SwiftStdlib/RangeReplaceableCollectionExtensions.swift:
```diff
@@ -174,4 +174,22 @@ public extension RangeReplaceableCollection {
+     - Parameter newElements: The optional element to append to the array
+     */
+    mutating func appendIfNonNil(_ newElement: Self.Element?) {
+            guard let newElement = newElement else {return}
+            self.append(newElement)
```
```suggestion
        guard let newElement = newElement else {return}
        self.append(newElement)
```

LucianoPAlmeida commented on Sources/SwifterSwift/SwiftStdlib/RangeReplaceableCollectionExtensions.swift:
```diff
@@ -174,4 +174,22 @@ public extension RangeReplaceableCollection {
+     - Parameter newElements: The optional sequence to append to the array
+     */
+    mutating func appendIfNonNil<S>(contentsOf newElements: S?) where Element == S.Element, S : Sequence {
+            guard let newElements = newElements else {return}
+            self.append(contentsOf: newElements)
```
```suggestion
       guard let newElements = newElements else {return}
       self.append(contentsOf: newElements)
```

jevonmao commented:
@LucianoPAlmeida Sorry about the formatting, i keep forgetting and messing  up which means more work for you. Thanks for the help.

LucianoPAlmeida commented on CHANGELOG.md:
```diff
@@ -60,7 +60,9 @@ The changelog for **SwifterSwift**. Also see the [releases](https://github.com/S
   - Added `.degrees(_:)`, `arcMinutes(_:)`, `arcSeconds(_:)`, `radians(_:)`, `gradians(_:)` and `revolutions(_:)`  to conveniently initialize measurement with corresponding unit. [#936](https://github.com/SwifterSwift/SwifterSwift/pull/936) by [Shiva Huang](https://github.com/ShivaHuang)
-
+- **Array**
+  - Added `append()` and `append(contentsOf` method overloads that can append optional element to `Array`. By [jevonmao](https://github.com/jevonmao)
```
One last thing can you update the new names and the fact that they are not overloads anymore and also not on `Array`

jevonmao replied:
Done.

LucianoPAlmeida commented on Sources/SwifterSwift/SwiftStdlib/RangeReplaceableCollectionExtensions.swift:
```diff
@@ -174,4 +174,22 @@ public extension RangeReplaceableCollection {
+     - Parameter newElements: The optional sequence to append to the array
+     */
+    mutating func appendIfNonNil<S>(contentsOf newElements: S?) where Element == S.Element, S : Sequence {
+       guard let newElements = newElements else {return}
+       self.append(contentsOf: newElements)
```
Sorry for bikesheding this formatting stuff, is really simple 
```suggestion
        guard let newElements = newElements else {return}
        self.append(contentsOf: newElements)
```

LucianoPAlmeida commented:
> @LucianoPAlmeida Sorry about the formatting, i keep forgetting and messing up which means more work for you. Thanks for the help.

No worries, and thank you for the contribution! I think we almost there =]

LucianoPAlmeida reviewed (approved):
Looks good, Thank you!

LucianoPAlmeida reviewed (approved):
This looks good to me, but I would like @guykogus opinion before merging

guykogus commented on CHANGELOG.md:
```diff
@@ -59,12 +59,12 @@ The changelog for **SwifterSwift**. Also see the [releases](https://github.com/S
   - Added `replacingOccurrences(ofPattern:withTemplate:options:searchRange:)` as a more convenient way to replace patterns. [#901](https://github.com/SwifterSwift/SwifterSwift/pull/901) by [gurgeous](https://github.com/gurgeous)
 - **Measurement**
-  - Added `.degrees(_:)`, `arcMinutes(_:)`, `arcSeconds(_:)`, `radians(_:)`, `gradians(_:)` and `revolutions(_:)`  to conveniently initialize measurement with corresponding unit. [#936](https://github.com/SwifterSwift/SwifterSwift/pull/936) by [Shiva Huang](https://github.com/ShivaHuang)
+  - Added `.degrees(_:)`, `arcMinutes(_:)`, `arcSeconds(_:)`, `radians(_:)`, `gradians(_:)` and `revolutions(_:)`  to conveniently initialize measurement with corresponding unit. [#936](https://github.com/SwifterSwift/SwifterSwift/pull/936) by [Shiva Huang](https://github.com/ShivaHuang)  
```
Why has this line been changed?
```suggestion
  - Added `.degrees(_:)`, `arcMinutes(_:)`, `arcSeconds(_:)`, `radians(_:)`, `gradians(_:)` and `revolutions(_:)`  to conveniently initialize measurement with corresponding unit. [#936](https://github.com/SwifterSwift/SwifterSwift/pull/936) by [Shiva Huang](https://github.com/ShivaHuang)
```

guykogus commented on CHANGELOG.md:
```diff
@@ -59,12 +59,12 @@ The changelog for **SwifterSwift**. Also see the [releases](https://github.com/S
 - **URL**
   - Added the `(unsafeString: String)` initializer for `URL` as a more conveniently to construct unsafe `URL`s from `String` by [jevonmao](https://github.com/jevonmao)
-
+  
```
Undo this too
```suggestion

```

guykogus commented on Sources/SwifterSwift/SwiftStdlib/RangeReplaceableCollectionExtensions.swift:
```diff
@@ -174,4 +174,22 @@ public extension RangeReplaceableCollection {
+    
+    /**
+     SwifterSwift: Adds a new element at the end of the array, mutates the array in place
+     - Parameter newElements: The optional element to append to the array
```
```suggestion
     - Parameter newElement: The optional element to append to the array
```

guykogus commented on Sources/SwifterSwift/SwiftStdlib/RangeReplaceableCollectionExtensions.swift:
```diff
@@ -174,4 +174,22 @@ public extension RangeReplaceableCollection {
+     - Parameter newElements: The optional element to append to the array
+     */
+    mutating func appendIfNonNil(_ newElement: Self.Element?) {
+        guard let newElement = newElement else {return}
```
```suggestion
        guard let newElement = newElement else { return }
```

guykogus commented on CHANGELOG.md:
```diff
@@ -134,6 +134,7 @@ The changelog for **SwifterSwift**. Also see the [releases](https://github.com/S
   - Added `init(grouping:by:)` to initialize a dictionary by grouping sequence from a hashable `KeyPath`. [#751](https://github.com/SwifterSwift/SwifterSwift/pull/751) by [mmdock](https://github.com/mmdock)
 - **RangeReplaceableCollection**:
   - Added `removeDuplicates(keyPath:)` for removing duplicate elements based on key path. [#737](https://github.com/SwifterSwift/SwifterSwift/pull/737) by [Ilya Glushchuk](https://github.com/iglushchuk).
+  - Added `appendIfNonNil()` and `appendIfNonNil(contentsOf: )` methods that can append optional elements and sequences, by [jevonmao](https://github.com/jevonmao)
```
```suggestion
  - Added `appendIfNonNil(_:)` and `appendIfNonNil(contentsOf:)` methods that can append optional elements and sequences, by [jevonmao](https://github.com/jevonmao)
```

guykogus commented on CHANGELOG.md:
```diff
@@ -134,6 +133,7 @@ The changelog for **SwifterSwift**. Also see the [releases](https://github.com/S
   - Added `init(grouping:by:)` to initialize a dictionary by grouping sequence from a hashable `KeyPath`. [#751](https://github.com/SwifterSwift/SwifterSwift/pull/751) by [mmdock](https://github.com/mmdock)
 - **RangeReplaceableCollection**:
   - Added `removeDuplicates(keyPath:)` for removing duplicate elements based on key path. [#737](https://github.com/SwifterSwift/SwifterSwift/pull/737) by [Ilya Glushchuk](https://github.com/iglushchuk).
+  - Added `appendIfNonNil(_:)` and `appendIfNonNil(contentsOf:)` methods that can append optional elements and sequences, by [jevonmao](https://github.com/jevonmao)
```
```suggestion
  - Added `appendIfNonNil(_:)` and `appendIfNonNil(contentsOf:)` methods that can append optional elements and sequences. [#966](https://github.com/SwifterSwift/SwifterSwift/pull/966)  by [jevonmao](https://github.com/jevonmao)
```

guykogus commented on Sources/SwifterSwift/SwiftStdlib/RangeReplaceableCollectionExtensions.swift:
```diff
@@ -174,4 +174,22 @@ public extension RangeReplaceableCollection {
+     SwifterSwift: Adds a new element at the end of the array, mutates the array in place
+     - Parameter newElement: The optional element to append to the array
+     */
+    mutating func appendIfNonNil(_ newElement: Self.Element?) {
```
```suggestion
    mutating func appendIfNonNil(_ newElement: Element?) {
```

LucianoPAlmeida commented on Sources/SwifterSwift/SwiftStdlib/RangeReplaceableCollectionExtensions.swift:
```diff
@@ -174,4 +174,22 @@ public extension RangeReplaceableCollection {
+     - Parameter newElements: The optional sequence to append to the array
+     */
+    mutating func appendIfNonNil<S>(contentsOf newElements: S?) where Element == S.Element, S : Sequence {
+        guard let newElements = newElements else {return}
```
Since there are some more minor changes to address, let's fix this also
```suggestion
        guard let newElements = newElements else { return }
```

guykogus commented on CHANGELOG.md:
```diff
@@ -136,6 +135,7 @@ The changelog for **SwifterSwift**. Also see the [releases](https://github.com/S
   - Added `init(grouping:by:)` to initialize a dictionary by grouping sequence from a hashable `KeyPath`. [#751](https://github.com/SwifterSwift/SwifterSwift/pull/751) by [mmdock](https://github.com/mmdock)
 - **RangeReplaceableCollection**:
   - Added `removeDuplicates(keyPath:)` for removing duplicate elements based on key path. [#737](https://github.com/SwifterSwift/SwifterSwift/pull/737) by [Ilya Glushchuk](https://github.com/iglushchuk).
+  - Added `appendIfNonNil(_:)` and `appendIfNonNil(contentsOf:)` methods that can append optional elements and sequences. [#966](https://github.com/SwifterSwift/SwifterSwift/pull/966)  by [jevonmao](https://github.com/jevonmao)
```
```suggestion
  - Added `appendIfNonNil(_:)` and `appendIfNonNil(contentsOf:)` methods that can append optional elements and sequences. [#966](https://github.com/SwifterSwift/SwifterSwift/pull/966) by [jevonmao](https://github.com/jevonmao)
```

guykogus reviewed (changes requested):
Sorry for being so nitpicky. It hurts to see the mistake 😅
